### PR TITLE
coredata: Correctly handle receiving a pipe for native/cross files

### LINKED ISSUE #5505: Meson 0.51 breaks Mesa build
Configuring Mesa (https://gitlab.freedesktop.org/mesa/mesa) with meson 0.51 causes `ninja`
to regenerate the build files in an infinite loop.
Mesa was configured like this, which works fine in earlier meson version:

```
meson "$build_dir" \
    --strip \
    -D buildtype=release \
    -D prefix="$install_dir" \
    -D b_lto=true \
    -D b_ndebug=true \
    -D platforms=drm,x11 \
    -D dri-drivers= \
    -D gallium-drivers= \
    -D vulkan-drivers=amd \
    -D gles1=false \
    -D gles2=false \
    -D opengl=false \
    -D valgrind=false
```

This was tested on Arch Linux, which just recently updated meson to 0.51.

### PR DESCRIPTION
**Pull request by dcbaker**
In some cases a cross/native file may be a pipe, such as when using bash
process replacement `meson --native-file
<([binaries]llvm-config='/opt/bin/llvm-config')`, for example. In this
case we copy the contents of the pipe into a file in the meson-private
directory so we can create a proper ninja dependency, and be able to
reload the file on --wipe/--reconfigure. This requires some extra
negotiation to preserve these native/cross files.

Fixes #5505

### CODE REVIEW

File: run_unittests.py [hunk 1]
```diff
@@ -5815,6 +5816,26 @@ def test_multiple_native_files_override(self):
             '--native-file', config, '--native-file', config2,
             '-Dcase=find_program'])
 
+    @unittest.skipIf(os.name != 'posix', 'Uses fifos, which are not available on non Unix OSes.')
+    def test_native_file_is_pipe(self):
+        fifo = os.path.join(self.builddir, 'native.file')
+        os.mkfifo(fifo)
+        with tempfile.TemporaryDirectory() as d:
+            wrapper = self.helper_create_binary_wrapper('bash', d, version='12345')
+            def filler():
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[Flake8]__ [E306] expected 1 blank line before a nested definition, found 0 ([link](https://sider.review/gh/repos/19784232/pulls/5523#issue-4249955))

<sub>posted by Sider</sub>

**Comment by nirbheek** [comment]
We didn't add this PR to the milestone so it didn't make it in the 0.51.1 release.

**Comment by dcbaker** [comment]
@nirbheek, so the PR and not just the issue needs to block the milestone? Just so I do the right thing in the future.

**Comment by nirbheek** [comment]
> @nirbheek, so the PR and not just the issue needs to block the milestone? Just so I do the right thing in the future.

Yep, the PR has to be in the milestone. Although, I guess my script should also check that the PRs associated with an issue are also in the milestone.

**Pull request merged**